thetis/callback.py

```diff
@@ -861,3 +870,151 @@ def message_str(self, *args):
+                self.variable_names.append(f_short)
+            else:
+                coords = ['x', 'y', 'z']
+                for k in range(self.field_dims[f]):
+                    f_comp_name = f_short + '_' + coords[k]
+                    self.variable_names.append(f_comp_name)
+
+        super().__init__(
+            solver_obj,
+            outputdir=outputdir,
+            array_dim=self.value_shape,
+            attrs=attrs,
+            export_to_hdf5=export_to_hdf5,
+            append_to_log=append_to_log)
+        self._initialized = False
+
+    def _initialize(self):
+        outputdir = self.outputdir
+        if outputdir is None:
+            outputdir = self.solver_obj.options.outputdir
+
+        # construct mesh points for evaluation
+        self.xy = list(zip(self.x, self.y))
+        self.trans_x = np.tile(self.x[np.newaxis, :], (self.n_points_z, 1))
+        self.trans_y = np.tile(self.y[np.newaxis, :], (self.n_points_z, 1))
+
+    def _update_coords(self):
+        try:
+            depth = np.array(self.solver_obj.fields.bathymetry_2d.at(self.xy))
+            elev = np.array(self.solver_obj.fields.elev_cg_2d.at(self.xy))
+        except PointNotInDomainError as e:
+            error('{:}: Transect "{:}" point out of horizontal domain'.format(self.__class__.__name__, self.location_name))
+            raise e
+        epsilon = 1e-5  # nudge points to avoid libspatialindex errors
+        z_min = -(depth - epsilon)
+        z_max = elev - epsilon
+        if self.force_z_min is not None:
+            z_min = np.maximum(z_min, self.force_z_min)
+        if self.force_z_max is not None:
+            z_max = np.minimum(z_max, self.force_z_max)
+        self.trans_z = np.linspace(z_max, z_min, self.n_points_z)
+        self.trans_z = self.trans_z.reshape(self.value_shape)
+        self.xyz = np.vstack((self.trans_x.ravel(),
+                              self.trans_y.ravel(),
+                              self.trans_z.ravel())).T
+
+    def __call__(self):
+        if not self._initialized:
+            self._initialize()
+        self._update_coords()
+
+        outvals = [self.trans_z]
+        for fieldname in self.fieldnames:
+            field = self.solver_obj.fields[fieldname]
+            field_dim = self.field_dims[fieldname]
+            try:
+                vals = field.at(tuple(self.xyz))
+                arr = np.array(vals)
+            except PointNotInDomainError as e:
+                error('{:}: Cannot evaluate data on transect {:}'.format(self.__class__.__name__, self.location_name))
+                raise e
+            shape = list(self.value_shape) + [field_dim]
+            arr = np.array(vals).reshape(shape)
+            if field_dim == 3:
+                outvals.extend([arr[..., 0], arr[..., 1], arr[..., 2]])
+            elif field_dim == 2:
+                outvals.extend([arr[..., 0], arr[..., 1]])
+            else:
+                outvals.extend([arr[..., 0]])
```

Comment:
I think this could be just one line:
```
outvals.extend(arr.T)
```

Reply:
`outvals` is actually a list in callback objects. But I can remove the if tree with
```python
            components = [arr[..., i] for i in range(arr.shape[-1])]
            outvals.extend(components)
```

Reply:
Yeah - actually it doesn't matter that it's a list, extend takes any iterable, so you can just iterate over `a.T` but on second thought that would actually flip the remaining dimensions (nxy and nz) as well which you probably don't want - so your solution looks good